2.6 Retrieve Unit Config From Model Cache

manadart opened the pull request:
## Description of change

This patch updates the uniter API to retrieve unit charm config settings from the model cache.

It is the first of two parts, the second of which will alter configuration _watching_ to work from the cache. It is broken up to ease reviewing because most of this patch ensures that the uniter test suite continues to work with the _cachetest_ package functionality mixed in.

It is known that watching the DB for changes and retrieving them from the cache is not a desirable state - the second patch will follow as quickly as can be managed.

The functionality change is a handful of lines in _apiserver/facades/agent/uniter/uniter.go_.

## QA steps

- Bootstrap
- `juju deploy redis`.
- `juju config redis password=new-pass`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "requirepass new-pass" is present.
- `juju config redis databases=4`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "databases 4" is present.
- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- `juju add-branch test-branch`.
- `juju track test-branch redis/0`.
- `juju config redis password=branch-pass --branch test-branch`.
- `juju config redis databases=8 --branch master` (the watcher is still watching settings in state, so we need to invoke it by changing master settings).
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that:
  - the line "databases 8" is present;
  - the line "requirepass branch-pass" (from branch-based settings) is present.

## Documentation changes

None.

## Bug reference

N/A

manadart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Get Unit Configuration From the Model Cache.](https://trello.com/c/W2oQDFcE/1566-get-unit-configuration-from-the-model-cache)

hmlanigan commented on apiserver/facades/agent/uniter/uniter_test.go:
```diff
@@ -1140,6 +1188,7 @@ func (s *uniterSuite) TestWatchActionNotifications(c *gc.C) {
 	wc.AssertNoChange()
 
 	addedAction, err := s.wordpressUnit.AddAction("fakeaction", nil)
+	c.Assert(err, jc.ErrorIsNil)
```
good catch

hmlanigan commented on apiserver/facades/agent/uniter/uniter_test.go:
```diff
@@ -713,13 +767,7 @@ func (s *uniterSuite) TestGetPrincipal(c *gc.C) {
-		LeadershipChecker_: s.State.LeadershipChecker(),
-	})
-	c.Assert(err, jc.ErrorIsNil)
+	subUniter := s.newUniterAPI(c, s.State, subAuthorizer)
```
Is not making this same change at line 251 deliberate?

manadart replied:
Yes; we are check the error there, rather than setting a working API.

hmlanigan commented on apiserver/facades/agent/uniter/uniter_test.go:
```diff
@@ -3753,15 +3779,7 @@ func (s *uniterNetworkInfoSuite) setupUniterAPIForUnit(c *gc.C, givenUnit *state
-		LeadershipChecker_: s.State.LeadershipChecker(),
-	})
-	c.Assert(err, jc.ErrorIsNil)
+	_ = s.newUniterAPI(c, s.State, s.authorizer)
```
The uniterNetworkInfoSuite doesn't have the mock ctrl setup, so this panics.

manadart commented on core/cache/cachetest/state.go:
```diff
@@ -130,16 +132,29 @@ func MachineChange(c *gc.C, modelUUID string, machine *state.Machine) cache.Mach
 func UnitChange(c *gc.C, modelUUID string, unit *state.Unit) cache.UnitChange {
 	publicAddr, err := unit.PublicAddress()
-	c.Assert(err, jc.ErrorIsNil)
+	if !network.IsNoAddressError(err) {
```
Done.

manadart commented:
Turns out the issue with `JujuConnSuite` is not that the cache is not populated as in a `StateSuite`, but that there needs to be some test synchronisation to ensure the cache is populated when we test cache-based functionality - `JujuConnSuite` sets up the cache-worker that populates the cache.

I have removed the explicit cache setting and added synchronisation points instead.

jameinel commented on apiserver/facades/agent/uniter/goal-state_test.go:
```diff
@@ -27,68 +25,24 @@ import (
-	mysqlUnit     *state.Unit
+	uniterSuiteBase
+
+	machine2 *state.Machine
```
it seems odd that we would strip out machine 0 and 1 but not 2. did we miss a step?

manadart replied:
It has `machine0` and `machine1` via the `uniterSuiteBase`. I just removed some repetition by using the base as intended.

jameinel commented on core/cache/cachetest/controller.go:
```diff
@@ -17,6 +17,9 @@ import (
 // without the need for a multi-watcher and cache worker.
+// This is useful when testing with StateSuite;
+// JujuConnSuite sets up a cache worker and multiwatcher
+// to keep the model cache in sync.
```
Is this comment still valid?
You sounded like you had taken a path that you changed your mind, but I don't see where 'controller.go' is doing this.

manadart replied:
I have clarified this comment to indicate that is is not necessary when using `JujuConnSuite`.

jameinel commented on core/cache/cachetest/controller.go:
```diff
@@ -95,6 +98,11 @@ func (tc *TestController) UpdateMachine(c *gc.C, modelUUID string, machine *stat
+// UpdateUnit updates the input state unit in the cache.
+func (tc *TestController) UpdateUnit(c *gc.C, modelUUID string, unit *state.Unit) {
+	tc.SendChange(UnitChange(c, modelUUID, unit))
+}
```
Nothing seems to call this function.

manadart replied:
Yes, I removed code that previously called it. I anticipate future use.

jameinel commented on core/cache/unit.go:
```diff
@@ -73,6 +73,10 @@ func (u *Unit) Ports() []network.Port {
 func (u *Unit) ConfigSettings() (charm.Settings, error) {
+	if u.details.CharmURL == "" {
+		return nil, errors.New("unit charm not set")
+	}
```
Is this indicating that we didn't get an event? Is this just a debugging prophylactic?

manadart replied:
This makes it congruent with behaviour of the state-based watcher. Existing unit tests assert this behaviour.

jameinel commented on apiserver/facades/agent/uniter/uniter.go:
```diff
@@ -1012,8 +1024,8 @@ func (u *UniterAPI) ConfigSettings(args params.Entities) (params.ConfigSettingsR
-			var unit *state.Unit
-			unit, err = u.getUnit(tag)
+			var unit cache.Unit
+			unit, err = u.getCacheUnit(tag)
```
I take it this is the core of the change and the rest is just dealing with the fallout?

manadart replied:
Yes.

manadart commented:
$$merge$$
